replica/distributed_loader.cc

```diff
@@ -664,16 +664,13 @@ future<> table_population_metadata::populate_subdir(sstring subdir, allow_offstr
     auto sstdir = get_path(subdir);
     dblog.debug("Populating {}/{}/{} allow_offstrategy_compaction={} must_exist={}", _ks, _cf, sstdir, do_allow_offstrategy_compaction, dir_must_exist);
 
-    if (!co_await file_exists(sstdir.native())) {
+    if (!_sstable_directories.contains(subdir)) {
         if (dir_must_exist) {
```

Comment:
> we're going to infer the state of a sstable from the subdirectory in its logical path, right?

No, but this is how things work now. There' a draft #12707 that tries to make location and state become independent up (or down) until the storage driver

Reply:
Thanks for the explanation. It seems to me that we could have decoupled logical from physical location (done transparely by underlying storage driver) and still infer the state from logical path. Filesystem driver would still have a 1:1 relationship.

Reply:
We can, but it's inconvenient for object storage driver. Consider it gets a "change state" request (nowadays -- move) asking to replace "foo/bar/staging" into "foo/bar". In the ownership table it cannot `UPDATE ... SET location = foo/bar WHERE localion = foo/bar/staging` because `location` is the primary key. So it would need to parse (well, split :roll_eyes: , but still) the path into "foo/bar" and { "staging" or "<empty>" } and then `UPDATE ... SET state = {new_state} WHERE location = foo/bar`

Reply:
In my idea, I assumed that s3 location doesn't depend on state, therefore the ownership table would point to the location in s3 bucket, and state would be a separate field (not member of primary key). When going from state A to B,  we'd only update that field. However, listing sstables in a given table per state would be hard. When opening a sstable, s3 driver knows how to compose the physical location (for talking to s3 storage) and logical path (to be used by sstable, etc).

Reply:
My idea and implementation is the same. But currently state is wired into sstable path and #12707 approaches to fix that. We'll be there eventually :)